addons/popochiu/migration/helpers/popochiu_migration_helper.gd

```diff
@@ -0,0 +1,256 @@
+		return PopochiuResources.GAME_PATH
+	else: # Error cannot access the game folders
+		return ""
+
+
+## Returns the user project migration version from the "res://game/popochiu_data.cfg" file.
+## If the Popochiu Migration Version is greater than the user project migration version
+## then a migration needs to be done.
+## If -1 gets returned then an error has occured.
+static func get_user_migration_version() -> int:
+	# popochiu_data.cfg config file could not be loaded, return error
+	if PopochiuResources.get_data_cfg() == null:
+		PopochiuUtils.print_error("Can't load [code]popochiu_data.cfg[/code] file.")
+		return -1
+	
+	if PopochiuResources.has_data_value("migration", "version"):
+		# Return the migration version in the popochiu_data.cfg file
+		return PopochiuResources.get_data_value("migration", "version", 1)
+	else:
+		# Run Migration 1 and so on
+		return 0
+
+
+## Returns [code]true[/code] if this is an empty project: no rooms, no characters, no inventory
+## items, no dialogues, and no audio files.
+static func is_empty_project() -> bool:
+	if (
+		get_game_path() == PopochiuResources.GAME_PATH
+		and PopochiuResources.get_section_keys("rooms").is_empty()
+		and PopochiuResources.get_section_keys("characters").is_empty()
+		and PopochiuResources.get_section_keys("inventory_items").is_empty()
+		and PopochiuResources.get_section_keys("dialogs").is_empty()
+		and PopochiuResources.get_section_keys("audio").is_empty()
+	):
+		return true
+	
+	return false
+
+
+## Returns [true] if the current Popochiu migration version is newer than the user's migration
+## version, which means a migration is needed.
+static func is_migration_needed() -> bool:
+	return get_migrations_count() > get_user_migration_version()
+
+
+## Updates [code]res://game/popochiu_data.cfg[/code] migration version to [param version].
+static func update_user_migration_version(new_version: int) -> void:
+	if PopochiuResources.set_data_value("migration", "version", new_version) != OK:
+		PopochiuUtils.print_error(
+			"Couldn't update the Migration version from [b]%d[/b] to [b]%d[/b] in Data file." % [
+				get_migrations_count(), new_version
+			]
+		)
+
+
+## Executes the [param steps] in [param migration] one by one and returns [code]true[/code] if
+## all finished without failing. It stops execution if a step fails.
+static func execute_migration_steps(migration: PopochiuMigration, steps: Array) -> bool:
+	if steps.is_empty():
+		PopochiuUtils.print_error(
+			"No steps to execute for Migration %d" % migration.get_version()
+		)
+		await PopochiuEditorHelper.wait_process_frame()
+		return false
+	
+	var idx := 0
+	for step: Callable in steps:
+		migration.start_step(idx)
+		var completion_type: PopochiuMigration.Completion = await step.call()
+		if completion_type in [
+			PopochiuMigration.Completion.DONE, PopochiuMigration.Completion.IGNORED
+		]:
+			await migration.step_finished(idx, completion_type)
+		else:
+			return false
+		
+		idx += 1
```

Comment:
I did this to avoid repeating code in the steps execution. Each time a step is executed, this code calls `start_step`, which is used by the popup to show a step started (a GIF shows it is in progress). When the step finishes, `step_finished` is called to mark the step as done (visually).

Reply:
If you think it's confusing or more complex than necessary, I can think of ways to refactor and simplify it.

Reply:
No no, I get it now and I like it.

```suggestion
		migration.start_step(idx)
		var completion_type: PopochiuMigration.Completion = await step.call()
		if completion_type in [
			PopochiuMigration.Completion.DONE, PopochiuMigration.Completion.IGNORED
		]:
			await migration.step_finished(idx, completion_type)
		else:
			return false
		
		idx += 1
```

Reply:
One thing that I notice thou is: if the completion_type is something we don't manage, you are returning false without stopping the step in the interface. Is there any case when you can have your interface stuck with an endless loader in case of errors? Or are you managing this out of this function?

Reply:
Hmmmm.... You're right. I'm not sure I have all the possible cases covered, but in most of them, if there is an error, the step will return `PopochiuMigration.Completion.FAILED`. I'll double-check it anyway.